enhancements/machine-api/converting-machine-api-to-cluster-api.md

```diff
@@ -0,0 +1,756 @@
+
+To ensure we always have the latest version, caches should not be used when performing this check (Cluster API is based on controller-runtime which by default uses a caching client).
+
+Each of the infrastructure Machine controllers will not have an equivalent resource to look up, and therefore must first identify the Machine that they are associated with and look up the equivalent resource for this Machine.
+
+The specifics of how this will be implemented are still to be determined.
+
+##### Filtering reconciles via a webhook
+
+An alternative to the above is to have each Cluster API controller perform an equivalent of a SubjectAccessReview.
+A request to a webhook with details of the resource they have been requested to reconcile, where the webhook can then check if the resource should be reconciled and return a response.
+When the response is denied, the reconcile should be aborted.
+
+If the webhook is unavailable, the mechanism should fail close and the reconcile should be retried at a later time.
+
+#### Filtering reconciles in Machine API controllers
+
+Each controller will, at the start of the reconcile logic, check for the presence of the authoritative annotation and will either skip or continue with the reconciliation based on the presence of this annotation.
+
+#### The Cluster Autoscaler in mixed MachineSet clusters
+
+The Cluster Autoscaler today looks at Machine API MachineSets and filters these based on a number of annotations to identify node groups that it may scale.
+In the future, where we have both Machine API and Cluster API authoritative MachineSets, the Autoscaler must be updated to ensure it scales the authoritative MachineSet.
+
+Since every Machine API MachineSet will be reflected into Cluster API, the autoscaler can be updated to reconcile only the Cluster API MachineSets.
+When a scaling decision is made, it must then update the authoritative MachineSet to the correct scale.
+
+To do this, there are two possible approaches to consider:
+- Look up the Machine API MachineSet mirror to determine the authoritative API and update accordingly
+  - This would be a carry patch to the autoscaler, which, may cause maintenance toil over time
+- Allow the Autoscaler to update the Cluster API MachineSet, but allow the synchronisation controller to identify the update has come from the Autoscaler, and mirror it to the authoritative MachineSet as appropriate
+  - This could be achieved by observing the managed fields property of the replicas field and determine which actor updated it.
+  If the Autoscaler last changed the replica count, it should be mirrored back to the Machine API MachineSet if it is authoritative.
+  - This creates an exception to the rules outline above and adds complication to the conversion, but, may be less of a burden than carrying a patch in the autoscaler
+
+Both avenues above will need to be explored further before a decision is made on how the Autoscaler will handle mixed MachineSet clusters.
```

Comment:
I think once we get started with the main conversion stuff it would be good to start getting the autoscaler ready, but until MAPI machines are being reflected into CAPI I don't think there's anything we can do there